node/core/dispute-coordinator/src/real/ordering.rs

```diff
@@ -0,0 +1,203 @@
+	included_candidates: HashSet<CandidateHash>,
+	/// including block -> `CandidateHash`
+	///
+	/// We need this to clean up `included_candidates` on `ActiveLeavesUpdate`.
+	candidates_by_block_number: BTreeMap<BlockNumber, HashSet<CandidateHash>>,
+}
+
+/// Comparator for ordering of disputes for candidates.
+///
+/// This comparator makes it possible to order disputes based on age and to ensure some fairness
+/// between chains in case of equally old disputes.
+///
+/// Objective ordering between nodes is important in case of lots disputes, so nodes will pull in
+/// the same direction and work on resolving the same disputes first. This ensures that we will
+/// conclude some disputes, even if there are lots of them. While any objective ordering would
+/// suffice for this goal, ordering by age ensures we are not only resolving disputes, but also
+/// resolve the oldest one first, which are also the most urgent and important ones to resolve.
+///
+/// Note: That by `oldest` we mean oldest in terms of relay chain block number, for any block
+/// number that has not yet been finalized. If a block has been finalized already it should be
+/// treated as low priority when it comes to disputes, as even in the case of a negative outcome,
+/// we are already too late. The ordering mechanism here serves to prevent this from happening in
+/// the first place.
+#[derive(Copy, Clone)]
+pub struct CandidateComparator {
+	/// Block number of the relay parent.
+	///
+	/// Important, so we will be participating in oldest disputes first.
+	///
+	/// Note: In fact it would make a bit more sense to use the `BlockNumber` of the including
+	/// block, as inclusion time is the actual relevant event when it comes to ordering. The
+	/// problem is, that a candidate can get included multiple times on forks, so the `BlockNumber`
+	/// of the including block is not unique. We could however easily work around that problem, by
+	/// just using the lowest `BlockNumber` of all available including blocks.
+	///
+	/// In practice it should not matter much, so in the interest of time I will not change the
+	/// implementation again now, but keep that in mind in case it becomes relevant at some point.
+	relay_parent_block_number: BlockNumber,
+	/// By adding the `CandidateHash`, we can guarantee a unique ordering across candidates.
+	candidate_hash: CandidateHash,
+}
+
+impl PartialEq for CandidateComparator {
+	fn eq(&self, other: &CandidateComparator) -> bool {
+		Ordering::Equal == self.cmp(other)
+	}
+}
+
+impl Eq for CandidateComparator {}
+
+impl PartialOrd for CandidateComparator {
+	fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+		Some(self.cmp(other))
+	}
+}
+
+impl Ord for CandidateComparator {
+	fn cmp(&self, other: &Self) -> Ordering {
+		match self.relay_parent_block_number.cmp(&other.relay_parent_block_number) {
+			Ordering::Equal => (),
+			o => return o,
+		}
+		self.candidate_hash.cmp(&other.candidate_hash)
+	}
+}
```

Comment:
why not derive this?

Reply:
Given that it is critical how the comparing is done - I preferred being explicit.